runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaTimerInternalsFactory.java

```diff
@@ -682,16 +653,26 @@ public void encode(TimerKey<K> value, OutputStream outStream)
         throws CoderException, IOException {
 
       // encode the timestamp first
-      STRING_CODER.encode(value.timerId, outStream);
-      STRING_CODER.encode(value.stateNamespace.stringKey(), outStream);
+      STRING_CODER.encode(value.getTimerId(), outStream);
+      STRING_CODER.encode(value.getStateNamespace().stringKey(), outStream);
 
       if (keyCoder != null) {
-        keyCoder.encode(value.key, outStream);
+        keyCoder.encode(value.getKey(), outStream);
+      }
+
+      // encode a byte (1/0) to indicate the presence/absence of timerFamilyId
+      // We can use this approach to add additional fields in the future
+      if (!Strings.isNullOrEmpty(value.getTimerFamilyId())) {
+        BOOLEAN_CODER.encode(true, outStream);
```

Comment:
could you help me understand why do we need a boolean flag? I suppose ```inStream.available() > 0``` could be sufficient to indicate there will be extra fields.

Reply:
This was just for future proofing in the event we add more fields to `TimerKey`. `TimerFamilyId` is optional and in the case it is null/empty, I wanted to encode a flag to indicate the absence of a timer family id. If we add more fields in the future, we can easily decode in a backward compatible way using the flag.

The only way we can probably omit encoding this boolean flag is to encode an empty string ("") whenever family id is not present.

Reply:
Got you, actually we do not need that because when timer family id is not present, we will still encode `null` in the bytes to actually indicate that timer family id is null.

Reply:
@ajothomas : I think we should be able to blindly encode timeFamilyId as it's going to be defaulted as blank "". See TimerInternals: line 198. The new TimerDataCoder (TimerDataCoderV2) also encode it directly without the need to do null check. We should be able to follow that.

Reply:
Thanks for catching this. As long as it defaults to empty string, `StringUtf8Coder` will be able to encode/decode it. Will remove the boolean flag.